### PR TITLE
ARROW-274: Add NullableMapVector to support nullable maps

### PR DESCRIPTION
**Pull request by julienledem**

### CODE REVIEW

**Comment by wesm** [comment]
Aside, has there been any more discussion in drill of removing the `NullableTYPEVector` distinction (i.e. having just one container type per vector)?

**Comment by julienledem** [comment]
@wesm For now, I'm fixing the gaps between the spec and the java implementation. Once that's done, we can refactor and remove unnecessary layers.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java [hunk 1]
```diff
@@ -126,16 +117,16 @@ public int getBufferSizeFor(final int valueCount) {
 
   @Override
   public ArrowBuf[] getBuffers(boolean clear) {
-    int expectedSize = getBufferSize();
-    int actualSize   = super.getBufferSize();
-
-    Preconditions.checkArgument(expectedSize == actualSize, expectedSize + " != " + actualSize);
+//    int expectedSize = getBufferSize();
```
Comment (hnfgns) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: Should we remove this method?

Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: will do.

**Comment by hnfgns** [comment]
+1. The patch looks good to me. Thanks for the works.

**Pull request closed without merging**